manim/utils/qhull.py

```diff
@@ -0,0 +1,163 @@
+#!/usr/bin/env python
+from __future__ import annotations
+
+import numpy as np
+
+
+class Point:
+    def __init__(self, coordinates):
+        self.coordinates = coordinates
+
+    def __hash__(self):
+        return hash(self.coordinates.tobytes())
+
+    def __eq__(self, other):
+        return np.array_equal(self.coordinates, other.coordinates)
+
+
+class SubFacet:
+    def __init__(self, coordinates):
+        self.coordinates = coordinates
+        self.points = frozenset(Point(c) for c in coordinates)
+
+    def __hash__(self):
+        return hash(self.points)
+
+    def __eq__(self, other):
+        return self.points == other.points
+
+
+class Facet:
+    def __init__(self, coordinates, normal=None, internal=None):
```

Comment:
There are two issues with the parameters:

- `internal`'s default value is `None`, but based on the current code, it can **never** be `None`: it is used in
```py
        self.normal = self.compute_normal(internal)
```
which does the following:
```py
        if np.dot(normal, self.center - internal) < 0:
```
which **will** crash if `internal` is `None`, with a `ValueError` because you can't subtract `None` from a `numpy.ndarray`.

I'd suggest not adding a default value to `internal`... but the name itself (`internal`) suggests to me that it might be a special, optional value, which kinda leads me to the next point:

- `normal` is currently not being used anywhere, and `Facet` calculates its `normal` attribute independently via `internal`. If this is intended, let's remove the `normal` parameter. But it seems to me that the intention could be to let an user add their own `normal`, and if they don't supply it, a `normal` would be calculated for them. Something like this:

```py
    if normal is None:
        normal = self.compute_normal(internal)
    self.normal = normal
```

in which case I can understand that `internal`'s default value might be `None`, if a user supplies `normal` instead. In that case I'd suggest:
```py
    if normal is None:
        if internal is None:
            raise ValueError("normal and internal can't be both None.")
        normal = self.compute_normal(internal)
    self.normal = normal
```

but it all depends on your intention with the code and the actual meaning of `internal` (is it an "internal value", or is it an internal something-else?). I'm currently just making wild guesses.

Reply:
Thank you so much for this (detailed) catch! This is an artifact from a previous version of the code where I computed the initial simplex's normals in batch and supplied them to `Facet` directly whereas every other use required an internal point to define the orientation. Since I don't expect users to interact with `Facet` in meaningful (do let me know if this is a reasonable expectation), I'll likely drop the `normal` parameter as you suggest.